
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>Release history &#8212; CPAT 1.2.1 documentation</title>
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="nav-item nav-item-0"><a href="#">CPAT 1.2.1 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="toctree-wrapper compound">
</div>
<a class="reference internal image-reference" href="_images/cpat.png"><img alt="alternate text" src="_images/cpat.png" style="width: 325.0px; height: 100.0px;" /></a>
<div class="section" id="release-history">
<h1>Release history<a class="headerlink" href="#release-history" title="Permalink to this headline">¶</a></h1>
<p><strong>CPAT v2.0.0</strong></p>
<blockquote>
<div><ol class="arabic simple">
<li><p>This version and all future versions will run in Python3. Installation use “pip3 install CPAT”</p></li>
</ol>
</div></blockquote>
<p><strong>CPAT v1.2.4</strong></p>
<blockquote>
<div><ol class="arabic simple">
<li><p>Supports installation using <a class="reference external" href="https://pypi.python.org/pypi/pip">pip</a></p></li>
<li><p>Fix inconsistence in LISCENSE.</p></li>
</ol>
</div></blockquote>
<p><strong>CPAT v1.2.3</strong></p>
<blockquote>
<div><p>v1.2.3 is almost identical to v1.2.2, with only one change:
In CPAT v1.2.2 (line 28), “import pysam” was unintentionally commented out. This error is fixed in v1.2.3.</p>
</div></blockquote>
<p><strong>CPAT v1.2.2</strong></p>
<blockquote>
<div><p>Fixed bugs in longest_orf() function. In CPAT (v1.2.2), if a mRNA sequence contains two ORFs with the same length,
it reports the one that is upstream (i.e. closer to transcript start site). While previous version reported the one
that detected first. For example:</p>
<p>This sequence contains two very simple open reading frames quoted by “</p>
<p>5’-A “ATGTAG” AA “ATGCTG”-3’</p>
<p>Old version will return “ATGCTG” simply because it uses the first reading frame, while current version will return “ATGTAG” because it’s closer to transcription start site.</p>
<p>Thanks Kristoffer Vitting-Seerup from University of Copenhagen for reporting this bug.</p>
</div></blockquote>
<p><strong>CPAT v1.2.1</strong></p>
<blockquote>
<div><ol class="arabic simple">
<li><p>Support compressed input file (<a href="#id1"><span class="problematic" id="id2">*</span></a>.gz or <a href="#id3"><span class="problematic" id="id4">*</span></a>.bz2)</p></li>
<li><p>Support url as input. url should be link (pointing to data repository) starts with <a class="reference external" href="http://">http://</a>, <a class="reference external" href="https://">https://</a> or <a class="reference external" href="ftp://">ftp://</a> .</p></li>
</ol>
</div></blockquote>
<p><strong>CPAT v1.2</strong></p>
<blockquote>
<div><p>Conservation score was obsolete. Because it depends on the alignment, relatively slow in
calculation and more importantly very little power is gained by using this feature. We
use hexamer usage bias as the 4th feature:</p>
<ol class="arabic simple">
<li><p>ORF (Open Reading Frame) size</p></li>
<li><p>ORF coverage (ratio of ORF size to transcript size)</p></li>
<li><p>Fickett TESTCODE statistic</p></li>
<li><p>Hexamer usage bias</p></li>
</ol>
</div></blockquote>
<p><strong>CPAT v1.1:</strong></p>
<blockquote>
<div><p>This is the only version using conservation score (phastCon) as one prediction feature. 4 features used for prediction:</p>
<ol class="arabic simple">
<li><p>ORF (Open Reading Frame) size</p></li>
<li><p>ORF coverage (ratio of ORF size to transcript size)</p></li>
<li><p>Fickett TESTCODE statistic</p></li>
<li><p>PhastCon conservation score</p></li>
</ol>
</div></blockquote>
</div>
<div class="section" id="introduction">
<h1>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h1>
<p>Using RNA-seq, tens of thousands of novel transcripts and isoforms have been identified (Djebali, et al  Nature, 2012
, Carbili et al, Gene &amp; Development, 2011)
The discovery of these hidden transcriptome rejuvenate the need of distinguishing coding
and noncoding RNA. However, Most previous coding potential prediction methods heavily rely
on alignment, either pairwise alignment to search for protein evidence or multiple alignments
to calculate phylogenetic conservation score (such as <a class="reference external" href="http://cpc.cbi.pku.edu.cn/">CPC</a> , <a class="reference external" href="http://compbio.mit.edu/PhyloCSF">PhyloCSF</a> and <a class="reference external" href="http://wash.github.com/rnacode/">RNACode</a> ). This is because most previously identified transcripts
including <strong>protein coding RNA</strong> and <strong>short, housekeeping/regulatory RNAs</strong> such as snRNAs,
snoRNA and tRNA are highly conserved. While still very useful, these approaches have several
limitations:</p>
<ol class="arabic simple">
<li><p>Most lncRNAs are less conserved and tend to be lineage specific which greatly limit the discrimination power of alignment-based methods. For example, of 550 lncRNAs detected from zebrafish, only 29 of them had detectable sequence similarity with putative mammalian orthologs (Ulitsky  et al, Cell, 2011).</p></li>
<li><p>A significant fraction of protein coding genes may have an alternatively processed isoform or one transcribed from an alternative promoter, these part of ncRNA cannot be correctly classified through homologous search because they would have significant match to protein coding genes.</p></li>
<li><p>Alignment based method is extremely slow. For example, <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/17631615">CPC</a> takes 6050 CPU minutes (&gt; 4 days) to evaluate 14,000 lncRNA transcripts.</p></li>
<li><p>Reliability depends on alignment quality. Most multi-alignment tools use heuristic search and do not guarantee to give optimal alignments.</p></li>
</ol>
<p>CPAT overcomes the above issues by using logistic regression model based on 4 pure sequence-based,
linguistic features</p>
<ol class="arabic simple">
<li><p>ORF size</p></li>
<li><p>ORF coverage</p></li>
<li><p><a class="reference external" href="http://nar.oxfordjournals.org/content/10/17/5303.abstract">Fickett TESTCODE</a></p></li>
<li><p><a class="reference external" href="http://nar.oxfordjournals.org/content/20/24/6441.abstract">Hexamer usage bias</a></p></li>
</ol>
<p>Linguistic features based method does not require other genomes or protein databases to perform alignment
and is more robust. Because it is alignment free, it runs much faster and also easier to use.
For example,  CPAT only took several minutes to evaluate the above 14,000 lncRNAs. More importantly,
compared with alignment-based approaches, CPAT achieves better sensitivity and
specificity (0.966 tested on human gene annotation).</p>
</div>
<div class="section" id="run-cpat-online">
<h1>Run <a class="reference external" href="http://lilab.research.bcm.edu/cpat/">CPAT online</a><a class="headerlink" href="#run-cpat-online" title="Permalink to this headline">¶</a></h1>
<p><a class="reference external" href="http://lilab.research.bcm.edu/cpat/">CPAT webserver</a></p>
<p><strong>Step1: Upload data to CPAT server. There are 3 different ways to uploada</strong></p>
<ul class="simple">
<li><p>Upload BED or FASTA format files from local disk. Files can be regular or compressed (<a href="#id6"><span class="problematic" id="id7">*</span></a>.gz, <a href="#id8"><span class="problematic" id="id9">*</span></a>.Z. <a href="#id10"><span class="problematic" id="id11">*</span></a>.z, <a href="#id12"><span class="problematic" id="id13">*</span></a>.bz, <a href="#id14"><span class="problematic" id="id15">*</span></a>.bz2, <a href="#id16"><span class="problematic" id="id17">*</span></a>.bzip2).</p></li>
<li><p>For small dataset, user can copy and paste data (in BED or FASTA format) directly to the text area.</p></li>
<li><p>For extremely larger dataset, user can save data in web server (http, https or ftp) first, then paste the data url to text area.</p></li>
</ul>
<p><strong>Step2: Select Select Species assembly</strong></p>
<p><strong>Step3: Click Submit button</strong></p>
<p><strong>NOTE</strong></p>
<ul class="simple">
<li><p>This web server only supports Human (hg19), Mouse (mm9 and mm10), Fly (dm3) and Zebrafish (Zv9).</p></li>
<li><p>When input file is BED format, the reference genome is required and the assembly version is important. For example, user cannot upload hg18 based BED file to this server, as we only support hg19.</p></li>
<li><p>When the input file is FASTA sequence file, the reference genome is NOT required, because all features can be calculated from the FASTA sequence directly.</p></li>
<li><p>FASTA sequence file is supported but not recommended, because if the file
size is too large, user may encounter ‘time out’ error. If you only have FASTA file, compressed
it using <em>gzip</em> or <em>bunzip2</em> before uploading. Or, you can save it to a file server and provide
the url to CPAT.</p></li>
</ul>
</div>
<div class="section" id="install-cpat-to-local-computer">
<h1>Install CPAT to local computer<a class="headerlink" href="#install-cpat-to-local-computer" title="Permalink to this headline">¶</a></h1>
<div class="section" id="download-cpat">
<h2>Download <a class="reference external" href="https://sourceforge.net/projects/rna-cpat/files/?source=navbar">CPAT</a><a class="headerlink" href="#download-cpat" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="https://sourceforge.net/projects/rna-cpat/files/?source=navbar">CPAT</a></p>
</div>
<div class="section" id="installation-for-cpat-version-1-2-4">
<h2>Installation for CPAT (version 1.2.4)<a class="headerlink" href="#installation-for-cpat-version-1-2-4" title="Permalink to this headline">¶</a></h2>
<p>Prerequisite: <a class="reference external" href="http://www.python.org/getit/releases/2.7/">python2.7</a>;  <a class="reference external" href="http://numpy.scipy.org/">numpy</a>; <a class="reference external" href="http://www.r-project.org/">R</a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span># install CPAT hosted on PyPI

$ pip2 install CPAT

#or you can download CPAT-VERSION.tar.gz first and then run:

$ pip2 install CPAT-VERSION.tar.gz
</pre></div>
</div>
</div>
<div class="section" id="installation-for-cpat-version-2-0-0-or-later-version">
<h2>Installation for CPAT (version 2.0.0 or later version)<a class="headerlink" href="#installation-for-cpat-version-2-0-0-or-later-version" title="Permalink to this headline">¶</a></h2>
<p>Prerequisite: <a class="reference external" href="https://www.python.org/download/releases/3.0/">python3.5</a> or later version;  <a class="reference external" href="http://numpy.scipy.org/">numpy</a>; <a class="reference external" href="http://www.r-project.org/">R</a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span># install CPAT hosted on PyPI

$ pip3 install CPAT

#or you can download CPAT-VERSION.tar.gz first and then run:

$ pip3 install CPAT-VERSION.tar.gz
</pre></div>
</div>
<p>NOTE:</p>
<ul class="simple">
<li><p>User need to download prebuilt <a class="reference external" href="https://sourceforge.net/projects/rna-cpat/files/v1.2.2/prebuilt_model/">logit model and hexamer table</a> for human, mouse, zebrafish and fly. For other species, we provide scripts to build these models (see below).</p></li>
<li><p>If you had “error: invalid command ‘egg_info’” when running ‘pip install CPAT-VERSION.tar.gz’. Try to upgrade setuptools:</p></li>
</ul>
<blockquote>
<div><p>$ pip install –upgrade setuptools</p>
</div></blockquote>
</div>
</div>
<div class="section" id="command-line-usage">
<h1>Command line usage<a class="headerlink" href="#command-line-usage" title="Permalink to this headline">¶</a></h1>
<div class="section" id="input-file">
<h2>Input file<a class="headerlink" href="#input-file" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><a class="reference external" href="http://dldcc-web.brc.bcm.edu/lilab/liguow/CGI/cpat/dat/human_test.bed">BED</a> format file (regular text or compressed). <a class="reference external" href="http://dldcc-web.brc.bcm.edu/lilab/liguow/CGI/cpat/dat/human_test.bed">BED</a> file should be in standard 12-column format.</p></li>
<li><p><a class="reference external" href="http://dldcc-web.brc.bcm.edu/lilab/liguow/CGI/cpat/dat/human_test.mRNA.fa">FASTA</a> format file (regular text or compressed)</p></li>
<li><p>a URL pointing to data that are saved remotely (data could be either BED or FASTA, either regular text or compressed file). <a class="reference external" href="http://">http://</a>, <a class="reference external" href="https://">https://</a> and <a class="reference external" href="ftp://">ftp://</a> are supported.</p></li>
</ul>
</div>
<div class="section" id="cpat-py">
<h2>cpat.py<a class="headerlink" href="#cpat-py" title="Permalink to this headline">¶</a></h2>
<p>User need to provide a gene file (‘-g’), a logit model file (‘-d’), a hexamer frequency table file (‘-x’)
and specify the output file name(‘-o’). Gene file could be either in <a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html#format1.7">BED</a>
or <a class="reference external" href="http://en.wikipedia.org/wiki/FASTA_format">FASTA</a> format. If in BED format, user also needs to
specify reference genome sequence (‘-r’); if gene file is in fasta format, each mRNA sequence must be in 5’-&gt;3’
direction.</p>
<dl>
<dt>Options:</dt><dd><dl class="option-list">
<dt><kbd><span class="option">--version</span></kbd></dt>
<dd><p>show program’s version number and exit</p>
</dd>
<dt><kbd><span class="option">-h</span>, <span class="option">--help</span></kbd></dt>
<dd><p>show this help message and exit</p>
</dd>
<dt><kbd><span class="option">-g <var>GENE_FILE</var></span>, <span class="option">--gene=<var>GENE_FILE</var></span></kbd></dt>
<dd><p>Transcripts either in BED format or mRNA sequences in
FASTA format: If this is BED format file, ‘-r’ must be
specified; if this is mRNA sequence file in FASTA
format, ignore the ‘-r’ option. The input BED or FASTA
file could be regular text file or compressed file
(<a href="#id24"><span class="problematic" id="id25">*</span></a>.gz, <a href="#id26"><span class="problematic" id="id27">*</span></a>.bz2) or accessible url.</p>
</dd>
<dt><kbd><span class="option">-o <var>OUT_FILE</var></span>, <span class="option">--outfile=<var>OUT_FILE</var></span></kbd></dt>
<dd><p>output file. Tab separated text file: geneID &lt;tab&gt;
mRNA size &lt;tab&gt; ORF size &lt;tab&gt; Fickett Score &lt;tab&gt;
Hexamer Score&lt;tab&gt;Coding Probability.</p>
</dd>
<dt><kbd><span class="option">-x <var>HEXAMER_DAT</var></span>, <span class="option">--hex=<var>HEXAMER_DAT</var></span></kbd></dt>
<dd><p>Prebuilt hexamer frequency table (Human, Mouse, Fly,
Zebrafish). Run ‘make_hexamer_tab.py’ to make this
table out of your own training dataset.</p>
</dd>
<dt><kbd><span class="option">-d <var>LOGIT_MODEL</var></span>, <span class="option">--logitModel=<var>LOGIT_MODEL</var></span></kbd></dt>
<dd><p>Prebuilt training model (Human, Mouse, Fly,
Zebrafish). Run ‘make_logitModel.py’ to build logit
model out of your own training datset</p>
</dd>
<dt><kbd><span class="option">-r <var>REF_GENOME</var></span>, <span class="option">--ref=<var>REF_GENOME</var></span></kbd></dt>
<dd><p>Reference genome sequences in FASTA format. Ignore
this option if mRNA sequences file was provided to
‘-g’. Reference genome file will be indexed
automatically (produce <a href="#id28"><span class="problematic" id="id29">*</span></a>.fai file along with the
original <a href="#id30"><span class="problematic" id="id31">*</span></a>.fa file within the same directory) if
hasn’t been done.</p>
</dd>
<dt><kbd><span class="option">-s <var>START_CODONS</var></span>, <span class="option">--start=<var>START_CODONS</var></span></kbd></dt>
<dd><p>Start codon (DNA sequence, so use ‘T’ instead of ‘U’)
used to define open reading frame (ORF). default=ATG</p>
</dd>
<dt><kbd><span class="option">-t <var>STOP_CODONS</var></span>, <span class="option">--stop=<var>STOP_CODONS</var></span></kbd></dt>
<dd><p>Stop codon (DNA sequence, so use ‘T’ instead of ‘U’)
used to define open reading frame (ORF). Multiple stop
codons should be separated by ‘,’. default=TAG,TAA,TGA</p>
</dd>
</dl>
</dd>
</dl>
<p>Examples 1:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cd CPAT-1.2.2/test

# use BED file as input. &#39;-r&#39; is required
$ cpat.py -r /database/hg19.fa -g Human_test_coding_mRNA_hg19.bed -d ../dat/Human_logitModel.RData -x ../dat/Human_Hexamer.tsv -o output1

# use FASTA file as input. &#39;-r&#39; is not required
$ cpat.py -g Human_test_coding_mRNA.fa  -d ../dat/Human_logitModel.RData -x ../dat/Human_Hexamer.tsv -o output2

$ head output1
mRNA_size      ORF_size        Fickett_score   Hexamer_score   coding_prob
NM_198317      2564    1929    1.1902  0.580124008851  0.999999999400781
NM_001014980   1043    909     1.1934  0.495124501575  0.999917485272474
NM_004421      2924    2013    1.0965  0.584852243126  0.999999999679495
NM_032348      2273    1329    1.2568  0.596113918225  0.999999649979388
NM_022834      4659    1338    1.1722  0.5589769999    0.999999395067735
NM_199121      4264    702     1.1722  0.560116277167  0.999276921874284
NM_001170535   2502    1761    1.2286  0.638700023819  0.999999997598248
NM_018188      2646    1905    1.2286  0.611789944749  0.999999999430874
NM_001170536   2332    1524    1.2712  0.661478359365  0.999999974636409


$ head output2
mRNA_size      ORF_size        Fickett_score   Hexamer_score   coding_prob
HG19_CT_USERTRACK_3545_NM_173083       3345    1677    0.7983  0.0216287267657 0.999998800005244
HG19_CT_USERTRACK_3545_NM_002744       2326    1779    1.2302  0.495437684259  0.999999995386405
HG19_CT_USERTRACK_3545_NM_001033581    2114    1230    1.3191  0.537935154638  0.999998793893276
HG19_CT_USERTRACK_3545_NM_001033582    2011    1230    1.3191  0.537935154638  0.999998800685703
HG19_CT_USERTRACK_3545_NM_032409       2660    1746    1.0164  0.361862240133  0.99999996805242
HG19_CT_USERTRACK_3545_NM_032236       4464    3108    0.9865  0.156660830868  1
HG19_CT_USERTRACK_3545_NM_130440       7706    5697    1.2799  0.579103525916  1
HG19_CT_USERTRACK_3545_NM_002840       7733    5724    1.2799  0.579964454487  1
HG19_CT_USERTRACK_3545_NM_007051       2583    1953    0.9702  0.184701969294  0.999999989213307
</pre></div>
</div>
</div>
<div class="section" id="make-hexamer-tab-py">
<h2>make_hexamer_tab.py<a class="headerlink" href="#make-hexamer-tab-py" title="Permalink to this headline">¶</a></h2>
<p><strong>make_hexamer_tab.py</strong> calculates the in frame hexamer (6mer) frequency from CDS sequence in fasta format. The CDS is mRNA sequence that removes UTR.
This table is required by CPAT to calculate the hexamer usage score. Users can download prebuilt hexamer tables (Human, Mouse, Fly, Zebrafish) from <a class="reference external" href="https://sourceforge.net/projects/rna-cpat/files/v1.2.2/prebuilt_model/">here</a></p>
<dl>
<dt>Options:</dt><dd><dl class="option-list">
<dt><kbd><span class="option">--version</span></kbd></dt>
<dd><p>show program’s version number and exit</p>
</dd>
<dt><kbd><span class="option">-h</span>, <span class="option">--help</span></kbd></dt>
<dd><p>show this help message and exit</p>
</dd>
<dt><kbd><span class="option">-c <var>CODING_FILE</var></span>, <span class="option">--cod=<var>CODING_FILE</var></span></kbd></dt>
<dd><p>Coding sequence (must be CDS without UTR, i.e. from start coden to stop coden) in fasta format.
User can get CDS sequence of a bed file using <a class="reference external" href="http://genome.ucsc.edu/cgi-bin/hgTables?org=Human&amp;db=hg19&amp;hgsid=289407045&amp;hgta_doMainPage=1">UCSC table browser</a></p>
</dd>
<dt><kbd><span class="option">-n <var>NONCODING_FILE</var></span>, <span class="option">--noncod=<var>NONCODING_FILE</var></span></kbd></dt>
<dd><p>Noncoding sequences in fasta format</p>
</dd>
</dl>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cd CPAT-1.2.2/test

$ make_hexamer_tab.py -c Human_coding_transcripts_CDS.fa   -n Human_noncoding_transcripts_RNA.fa &gt;Human_Hexamer.tsv

$ head Human_Hexamer.tsv

hexamer        coding  noncoding
GAACGT 0.000114999540425       6.20287252729e-05
CTTCTT 0.000280298143192       0.000464526231488
CACCCT 0.000254883880114       0.000337895737524
GAACGG 0.000178535198119       5.8077265737e-05
GAACGC 0.000136389878516       6.03746259323e-05
GAACGA 0.00015830968042        5.87205265917e-05
CACCCA 0.000258696019576       0.000448628498937
CTTCTA 0.000147508618612       0.000280645521457
CACCCC 0.000328479350276       0.000342582352322
...
</pre></div>
</div>
</div>
<div class="section" id="make-logitmodel-py">
<h2>make_logitModel.py<a class="headerlink" href="#make-logitmodel-py" title="Permalink to this headline">¶</a></h2>
<p>Build logistic regression model (“prefix.logit.RData”) required by CPAT. This program will output
3 files:</p>
<ul class="simple">
<li><p>prefix.feature.xls: A table contains features calculated from training datasets (coding and noncoding gene lists).</p></li>
<li><p>prefix.logit.RData: logit model required by CPAT (if R was installed).</p></li>
<li><p>prefix.make_logitModel.r: R script to build the above logit model.</p></li>
</ul>
<p>Note: Users can download prebuilt logit models (Human, Mouse, Fly, Zebrafish) from <a class="reference external" href="https://sourceforge.net/projects/rna-cpat/files/v1.2.2/prebuilt_model">here</a></p>
<dl>
<dt>Options:</dt><dd><dl class="option-list">
<dt><kbd><span class="option">--version</span></kbd></dt>
<dd><p>show program’s version number and exit</p>
</dd>
<dt><kbd><span class="option">-h</span>, <span class="option">--help</span></kbd></dt>
<dd><p>show this help message and exit</p>
</dd>
<dt><kbd><span class="option">-c <var>CODING_FILE</var></span>, <span class="option">--cgene=<var>CODING_FILE</var></span></kbd></dt>
<dd><p>Protein coding transcripts (used to build logit model)
either in BED format or mRNA sequences in FASTA
format: If this is BED format file, ‘-r’ must be
specified; if this is mRNA sequence file in FASTA
format, ignore the ‘-r’ option. The input BED or FASTA
file could be regular text file or compressed file
(<a href="#id33"><span class="problematic" id="id34">*</span></a>.gz, <a href="#id35"><span class="problematic" id="id36">*</span></a>.bz2) or accessible url. NOTE: transcript ID
should be unique.</p>
</dd>
<dt><kbd><span class="option">-n <var>NONCODING_FILE</var></span>, <span class="option">--ngene=<var>NONCODING_FILE</var></span></kbd></dt>
<dd><p>Non protein coding transcripts (used to build logit
model) either in BED format or mRNA sequences in FASTA
format: If this is BED format file, ‘-r’ must be
specified; if this is mRNA sequence file in FASTA
format, ignore the ‘-r’ option. The input BED or FASTA
file could be regular text file or compressed file
(<a href="#id37"><span class="problematic" id="id38">*</span></a>.gz, <a href="#id39"><span class="problematic" id="id40">*</span></a>.bz2) or accessible url.  NOTE: transcript ID
should be unique.</p>
</dd>
<dt><kbd><span class="option">-o <var>OUT_FILE</var></span>, <span class="option">--outfile=<var>OUT_FILE</var></span></kbd></dt>
<dd><p>output prefix.</p>
</dd>
<dt><kbd><span class="option">-x <var>HEXAMER_DAT</var></span>, <span class="option">--hex=<var>HEXAMER_DAT</var></span></kbd></dt>
<dd><p>Prebuilt hexamer frequency table (Human, Mouse, Fly,
Zebrafish). Run ‘make_hexamer_tab.py’ to generate this
table.</p>
</dd>
<dt><kbd><span class="option">-r <var>REF_GENOME</var></span>, <span class="option">--ref=<var>REF_GENOME</var></span></kbd></dt>
<dd><p>Reference genome sequences in FASTA format. Ignore
this option if mRNA sequences file was provided to
‘-g’. Reference genome file will be indexed
automatically (produce <a href="#id41"><span class="problematic" id="id42">*</span></a>.fai file along with the
original <a href="#id43"><span class="problematic" id="id44">*</span></a>.fa file within the same directory) if
hasn’t been done.</p>
</dd>
<dt><kbd><span class="option">-s <var>START_CODONS</var></span>, <span class="option">--start=<var>START_CODONS</var></span></kbd></dt>
<dd><p>Start codon (DNA sequence, so use ‘T’ instead of ‘U’)
used to define open reading frame (ORF). default=ATG</p>
</dd>
<dt><kbd><span class="option">-t <var>STOP_CODONS</var></span>, <span class="option">--stop=<var>STOP_CODONS</var></span></kbd></dt>
<dd><p>Stop codon (DNA sequence, so use ‘T’ instead of ‘U’)
used to define open reading frame (ORF). Multiple stop
codons should be separated by ‘,’. default=TAG,TAA,TGA</p>
</dd>
</dl>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cd CPAT-1.2.2/test

$ make_logitModel.py  -x Human_Hexamer.tsv -c Human_coding_transcripts_mRNA.fa -n Human_noncoding_transcripts_RNA.fa -o Human

Process protein coding transcripts: Human_coding_transcripts_mRNA.fa
Input gene file is in FASTA format
Process non coding transcripts: Human_noncoding_transcripts_RNA.fa
Input gene file is in FASTA format
build logi model ...
Warning message:
glm.fit: fitted probabilities numerically 0 or 1 occurred

#or use BED file as input
$ make_logitModel.py  -x Human_Hexamer.tsv -c Human_coding_transcripts_hg19.bed -n Human_noncoding_transcripts_hg19.bed  -r /database/hg19.fa  -o Human
</pre></div>
</div>
</div>
</div>
<div class="section" id="how-to-choose-cutoff">
<h1>How to choose cutoff<a class="headerlink" href="#how-to-choose-cutoff" title="Permalink to this headline">¶</a></h1>
<p><strong>Optimum cutoff were determined from TG-ROC</strong>.</p>
<ul class="simple">
<li><p>Human coding probability (CP) cutoff: 0.364 (CP &gt;=0.364 indicates coding sequence, CP &lt; 0.364 indicates noncoding sequence) (see performance figure D)</p></li>
<li><p>Mouse coding probability (CP) cutoff: 0.44</p></li>
<li><p>Fly coding probability (CP) cutoff: 0.39</p></li>
<li><p>Zebrafish coding probability (CP) cutoff: 0.38</p></li>
</ul>
<p><strong>Here we provide the R code and the data that we used to generate</strong> <a class="reference external" href="http://nar.oxfordjournals.org/content/41/6/e74/F3.expansion.html">Figure 3</a> <strong>in our paper</strong>. <strong>Note the</strong> <a class="reference external" href="http://cran.r-project.org/web/packages/ROCR/index.html">ROCR</a> <strong>library is required to run our R code</strong>.</p>
<p>1) Download R code and data from <a class="reference external" href="https://sourceforge.net/projects/rna-cpat/files/Figure3_data/">here</a></p>
<p>2) Put the R code and the data table in the same folder</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ ls
10Fold_CrossValidation.r       Human_train.dat
</pre></div>
</div>
<p>3) Run the R code from command line or console. The R code will perform 10-fold cross validation and generate Figure_3.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ Rscript 10Fold_CrossValidation.r     # install ROCR before running this code

Loading required package: gplots
Attaching package: ‘gplots’
The following object is masked from ‘package:stats’:

    lowess

Loading required package: methods
Warning message:
package ‘gplots’ was built under R version 3.1.2
[1] &quot;ID&quot;      &quot;mRNA&quot;    &quot;ORF&quot;     &quot;Fickett&quot; &quot;Hexamer&quot; &quot;Label&quot;
Warning message:
glm.fit: fitted probabilities numerically 0 or 1 occurred
Warning message:
glm.fit: fitted probabilities numerically 0 or 1 occurred
Warning message:
glm.fit: fitted probabilities numerically 0 or 1 occurred
Warning message:
glm.fit: fitted probabilities numerically 0 or 1 occurred
Warning message:
glm.fit: fitted probabilities numerically 0 or 1 occurred
Warning message:
glm.fit: fitted probabilities numerically 0 or 1 occurred
Warning message:
glm.fit: fitted probabilities numerically 0 or 1 occurred
Warning message:
glm.fit: fitted probabilities numerically 0 or 1 occurred
Warning message:
glm.fit: fitted probabilities numerically 0 or 1 occurred
Warning message:
glm.fit: fitted probabilities numerically 0 or 1 occurred
null device
          1
</pre></div>
</div>
</div>
<div class="section" id="how-to-prepare-training-dataset">
<h1>How to prepare training dataset<a class="headerlink" href="#how-to-prepare-training-dataset" title="Permalink to this headline">¶</a></h1>
<p>We prebuild hexamer tables and logit models for <a class="reference external" href="https://sourceforge.net/projects/rna-cpat/files/v1.2.2/prebuilt_model/">human, mouse, fly and zebrafish</a>.
If you want to run CPAT for other species, you need to prepare “coding sequences” and “noncoding sequences” as training data. These two files are required when you run <em>make_hexamer_tab.py</em> and <em>make_logitModel.py</em>.</p>
<ul class="simple">
<li><p><strong>Coding sequences</strong>: the whole CDS part of the mRNA. In other words, each CDS sequences should start with <strong>start codon</strong> and end with <strong>stop coden</strong>. As you expected, the length of each CDS should be <strong>an integer multiple of 3</strong>. You should reverse-compliment the CDS sequence if the mRNA sequence is extracted from  the genome and the gene is located on ‘-‘ strand.</p></li>
<li><p><strong>Noncoding sequences</strong>: It’s better to use those <strong>annotated (known) noncoding genes</strong>, rather than the “noncoding parts” of protein coding genes such as 3’UTR and 5’UTR.</p></li>
<li><p>It’s better to have balanced training dataset (i.e. the number of coding sequences is roughly equal to the number of noncoding sequences).</p></li>
<li><p>If the genome of the species you are working on is NOT well annotated and does not have enough “coding” and “noncoding” genes to build the training data, you could build your model using data from other species that is evolutionary close to the species you are working on.</p></li>
</ul>
</div>
<div class="section" id="evaluating-performance">
<h1>Evaluating Performance<a class="headerlink" href="#evaluating-performance" title="Permalink to this headline">¶</a></h1>
<p>Combinatorial effects of 3 major features. 10,000 coding genes (red dots) and 10,000 noncoding genes (blue dots) are clearly separated into two clusters. (below figure)</p>
<div class="figure align-default" style="width: 0.90px">
<a class="reference internal image-reference" href="_images/Figure_1A_features_3D.png"><img alt="_images/Figure_1A_features_3D.png" src="_images/Figure_1A_features_3D.png" style="width: 600.0px; height: 600.0px;" /></a>
</div>
<p>Performance evaluation using 10-fold cross validation (10,000 coding genes and 10,000 noncoding genes). Blue dotted curves represent the
10-fold cross validations, red solid curve represents the averaged curve between 10 runs of
validations. (A) ROC curve. (B) PR (precision-recall) curve. (C) Accuracy vs cutoff value.
(D) Two graphic ROC curve to determine the optimum cutoff value.</p>
<div class="figure align-default" style="width: 0.90px">
<a class="reference internal image-reference" href="_images/CPAT_performance.png"><img alt="_images/CPAT_performance.png" src="_images/CPAT_performance.png" style="width: 600.0px; height: 600.0px;" /></a>
</div>
</div>
<div class="section" id="comparison">
<h1>Comparison<a class="headerlink" href="#comparison" title="Permalink to this headline">¶</a></h1>
<p>To compare CPAT with CPC and PhyloCSF, we build an independent testing dataset that composed
of 4,000 high quality protein coding genes from Refseq annotation and 4,000 lincRNAs from
Human lincRNA catalog (Cabili et al., 2011). All 8000 genes were not included in the training
dataset of CPAT.</p>
<div class="figure align-default" style="width: 0.90px">
<a class="reference internal image-reference" href="_images/Figure_4.png"><img alt="_images/Figure_4.png" src="_images/Figure_4.png" style="width: 600.0px; height: 600.0px;" /></a>
</div>
<div class="figure align-default" style="width: 0.90px">
<a class="reference internal image-reference" href="_images/Figure_S2.png"><img alt="_images/Figure_S2.png" src="_images/Figure_S2.png" style="width: 600.0px; height: 600.0px;" /></a>
</div>
</div>
<div class="section" id="license">
<h1>LICENSE<a class="headerlink" href="#license" title="Permalink to this headline">¶</a></h1>
<p>CPAT is distributed under <a class="reference external" href="http://www.gnu.org/copyleft/gpl.html">GNU General Public License</a></p>
<p>This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License as
published by the Free Software Foundation; either version 2 of the
License, or (at your option) any later version. This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details. You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
02110-1301 USA</p>
</div>
<div class="section" id="reference">
<h1>Reference<a class="headerlink" href="#reference" title="Permalink to this headline">¶</a></h1>
<p>Wang, L., Park, H. J., Dasari, S., Wang, S., Kocher, J.-P., &amp; Li, W. (2013). CPAT: Coding-Potential Assessment Tool using an alignment-free logistic regression model. Nucleic Acids Research, 41(6), e74. <a class="reference external" href="http://nar.oxfordjournals.org/content/41/6/e74">doi:10.1093/nar/gkt006</a></p>
</div>
<div class="section" id="contact">
<h1>Contact<a class="headerlink" href="#contact" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><p>Liguo Wang: wang.liguo AT mayo.edu</p></li>
<li><p>Hyun Jung Park: hjpark AT bcm.edu</p></li>
</ul>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="#">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Release history</a></li>
<li><a class="reference internal" href="#introduction">Introduction</a></li>
<li><a class="reference internal" href="#run-cpat-online">Run CPAT online</a></li>
<li><a class="reference internal" href="#install-cpat-to-local-computer">Install CPAT to local computer</a><ul>
<li><a class="reference internal" href="#download-cpat">Download CPAT</a></li>
<li><a class="reference internal" href="#installation-for-cpat-version-1-2-4">Installation for CPAT (version 1.2.4)</a></li>
<li><a class="reference internal" href="#installation-for-cpat-version-2-0-0-or-later-version">Installation for CPAT (version 2.0.0 or later version)</a></li>
</ul>
</li>
<li><a class="reference internal" href="#command-line-usage">Command line usage</a><ul>
<li><a class="reference internal" href="#input-file">Input file</a></li>
<li><a class="reference internal" href="#cpat-py">cpat.py</a></li>
<li><a class="reference internal" href="#make-hexamer-tab-py">make_hexamer_tab.py</a></li>
<li><a class="reference internal" href="#make-logitmodel-py">make_logitModel.py</a></li>
</ul>
</li>
<li><a class="reference internal" href="#how-to-choose-cutoff">How to choose cutoff</a></li>
<li><a class="reference internal" href="#how-to-prepare-training-dataset">How to prepare training dataset</a></li>
<li><a class="reference internal" href="#evaluating-performance">Evaluating Performance</a></li>
<li><a class="reference internal" href="#comparison">Comparison</a></li>
<li><a class="reference internal" href="#license">LICENSE</a></li>
<li><a class="reference internal" href="#reference">Reference</a></li>
<li><a class="reference internal" href="#contact">Contact</a></li>
</ul>

  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/index.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="nav-item nav-item-0"><a href="#">CPAT 1.2.1 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2012, Liguo Wang, Hyun Jung Park.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.1.2.
    </div>
  </body>
</html>